src/org/opensolaris/opengrok/configuration/RuntimeEnvironment.java

```diff
@@ -1100,4 +1217,91 @@ public void run() {
 
         return ret;
     }
+
+    private Thread watchDogThread;
+    private WatchService watchDogWatcher;
+
+    /**
+     * Starts a watch dog service for a directory. It automatically reloads the
+     * AuthorizationFramework if there was a change.
+     *
+     * You can control start of this service by context-parameter in web.xml
+     * param-name: enableAuthorizationWatchDog
+     *
+     * @param directory root directory for plugins
+     */
+    public void startWatchDogService(File directory) {
+        if (directory == null || !directory.isDirectory() || !directory.canRead()) {
+            LOGGER.log(Level.INFO, "Watch dog cannot be started - invalid directory: {0}", directory);
+            return;
+        }
+        watchDogThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    watchDogWatcher = FileSystems.getDefault().newWatchService();
+                    Path dir = Paths.get(directory.getAbsolutePath());
+
+                    Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+                        @Override
+                        public FileVisitResult postVisitDirectory(Path d, IOException exc) throws IOException {
+                            // attach monitor
+                            d.register(watchDogWatcher, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+                            return CONTINUE;
+                        }
+                    });
+
+                    while (true) {
+                        final WatchKey key;
+                        try {
+                            key = watchDogWatcher.take();
+                        } catch (ClosedWatchServiceException x) {
+                            break;
+                        }
+                        boolean reload = false;
+                        for (WatchEvent<?> event : key.pollEvents()) {
+                            final WatchEvent.Kind<?> kind = event.kind();
+
+                            if (kind == ENTRY_CREATE) {
+                                reload = true;
+                            } else if (kind == ENTRY_DELETE) {
+                                reload = true;
+                            } else if (kind == ENTRY_MODIFY) {
+                                reload = true;
+                            }
+                        }
+                        if (reload) {
+                            Thread.sleep(2000); // experimental wait if file is being written right now
```

Comment:
I extracted this as a constant in the class context.
I wouldn't make it a param because it
 - should be enough always
 - is designed for development only; so anyone can change it while developing plugins

The motivation was just that the modify event is fired when the modification begins (for example compiling) and then framework can load an incomplete class).